### ISSUE TITLE
postponed regular expression counts wrong

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6271]
Migrated from [rt.perl.org#20684](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20684) (status was 'resolved')

Searchable as RT20684$

**Comment by p5pRT** [comment]
### From perl-5.8.0@ton.iguana.be

#### Created by perl-5.8.0@ton.iguana.be



perl \-wle '$z="A"; $\_="QDADEBCAV"; print "Did "\, s/\(??\{~~$z\}\)/$z\+\+;"\*"/eg\, " subs"; print'
Did 1 subs
QD\*DE\*\*AV

As you can see\, it however did 3 substitutions\, not 1

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.8.0:

Configured by ton at Tue Nov 12 01:56:18 CET 2002.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.19, archname=i686-linux-thread-multi-64int-ld
    uname='linux quasar 2.4.19 #5 wed oct 2 02:34:25 cest 2002 i686 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=define use64bitall=undef uselongdouble=define
    usemymalloc=y, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -fomit-frame-pointer',
    cppflags='-D_REENTRANT -D_GNU_SOURCE -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010315 (release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long long', ivsize=8, nvtype='long double', nvsize=12, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -ldb -ldl -lm -lpthread -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lpthread -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.2.4.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.2.4'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /usr/lib/perl5/5.8.0/i686-linux-thread-multi-64int-ld
    /usr/lib/perl5/5.8.0
    /usr/lib/perl5/site_perl/5.8.0/i686-linux-thread-multi-64int-ld
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.8.0:
    HOME=/home/ton
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/ton/bin.Linux:/home/ton/bin:/home/ton/bin.SampleSetup:/usr/local/bin:/usr/local/sbin:/usr/local/jre/bin:/home/oracle/product/9.0.1/bin:/usr/local/ar/bin:/usr/games/bin:/usr/X11R6/bin:/usr/share/bin:/usr/bin:/usr/sbin:/bin:/sbin:.
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

$ perl \-le '$\_="111"; print s/\(??\{1\}\)/2/ge'

This should print 3 but prints 0

Patch&#8203;:


---
<details><summary>Inline Patch</summary>

```diff
--- /arc/perl-current/pp_ctl.c	2003-02-16 16:24:51.000000000 +0200
+++ perl-current/pp_ctl.c	2003-02-21 08:40:36.000000000 +0200
@@ -200,7 +200,7 @@ PP(pp_substcont)
 	    sv_free(dstr);
 
 	    TAINT_IF(cx->sb_rxtainted & 1);
-	    PUSHs(sv_2mortal(newSViv((I32)cx->sb_iters - 1)));
+	    PUSHs(sv_2mortal(newSViv(saviters - 1)));
 
 	    (void)SvPOK_only_UTF8(targ);
 	    TAINT_IF(cx->sb_rxtainted);
-------------------------------------------------------------------

```
</details>
#! /usr/bin/perl
require "test\.pl";
\{
&nbsp;    $\_="1111"; is\(s/\(??\{1\}\)/x/eg\,4\,'\#20684 s/// with \(??\{\.\.\}\) inside'\);
\}
\_\_END\_\_


---
Of course\, it would be better to do all inside S\_regtry\, but that's
a little bit trickier\. Since pp\_substcont was already patched to
save then restore cxstack\->sb\_iters \( \< 5\.8\.0 \) \, I used that\.

Regards

Adi

**Comment by p5pRT** [comment]
### From @hvds

Enache Adrian \<enache@&#8203;rdslink\.ro> wrote&#8203;:
:$ perl \-le '$\_="111"; print s/\(??\{1\}\)/2/ge'
:
:This should print 3 but prints 0

Thanks\, applied as \#18775\.

:    $\_="1111"; is\(s/\(??\{1\}\)/x/eg\,4\,'\#20684 s/// with \(??\{\.\.\}\) inside'\);

This tries to eval "x"\, so I changed it back to 2\.

Hugo

**Comment by p5pRT** [comment]
### From enache@rdslink.ro

On Tue\, Feb 25\, 2003 at 03&#8203;:05&#8203;:42PM \+0000\, hv@&#8203;crypt\.org wrote&#8203;:
> Thanks\, applied as \#18775\.

It seems to fix \[perl \#7804\] too\.

Adi

\-\-\- from \#7804 \-\-\-
$\_ = 'abc';
$s = s\{ a \(?\{ 'A' \}\) | b \(?\{ 'B' \}\) \}\{$^R\}xg;
print "\<$\_> \<$s>\\n";

outputs : \(with perl 5\.7\.2\)
\<ABc> \<0>
But this should output \<ABc> \<2> instead\.

**Comment by p5pRT** [comment]
@cwest - Status changed from 'new' to 'resolved'
